dump operator names of a module and its sub-modules.

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29208 dump operator names of a module and its sub-modules.**

A binary to dump operator names from a script model and its sub models. 
Usage:
dump_operator_names path/to/script_model.pt path/to/output.yaml

Differential Revision: [D18350353](https://our.internmc.facebook.com/intern/diff/D18350353)

Commit:
dump operator names of a module and its sub-modules.

Commit:
Update on "dump operator names of a module and its sub-modules."

Commit:
Update on "dump operator names of a module and its sub-modules."

Commit:
Update on "dump operator names of a module and its sub-modules."

ljk53 commented on binaries/dump_operator_names.cc:
```diff
@@ -0,0 +1,54 @@
+    }
+  }
+  auto modules = m.get_modules();
+  for (const auto& sub_m : modules) {
```
Can sub-module have sub-module too?

iseeyuan replied:
Yes. It's a recursive call (line 35) and should cover all levels of submodules.

ljk53 commented on binaries/dump_operator_names.cc:
```diff
@@ -0,0 +1,54 @@
+    std::cerr << "usage: <path-to-script-module> <path-to-output-yaml>\n";
+    return 1;
+  }
+  auto m = torch::jit::load(argv[1]);
```
nit: can you use C10_DEFINE_string to define & parse command line flags, like other binaries/*.cpp files?

ljk53 commented on binaries/dump_operator_names.cc:
```diff
@@ -0,0 +1,54 @@
+void dump_opnames(const script::Module& m, std::unordered_set<std::string>& opnames) {
+  auto methods = m.get_methods();
+  for (const auto &method : methods) {
+    const auto &func = method.function();
```
nit: "const auto &" -> "const auto&", to be consistent with other places in this file.

ljk53 replied:
should update both line 16 and 17?

Commit:
Update on "dump operator names of a module and its sub-modules."


A binary to dump operator names from a script model and its sub models. 
Usage:
dump_operator_names path/to/script_model.pt path/to/output.yaml

Commit:
Update on "dump operator names of a module and its sub-modules."


A binary to dump operator names from a script model and its sub models. 
Usage:
dump_operator_names path/to/script_model.pt path/to/output.yaml

ljk53 commented on binaries/dump_operator_names.cc:
```diff
@@ -0,0 +1,69 @@
+        auto opname = node->schema().operator_name();
+        std::string namestr = opname.name;
+        if (!opname.overload_name.empty())
+          namestr += "." + opname.overload_name;
```
is overload_name being used? the codegen script will take the output of this binary as a whitelist of root ops that need to be kept. If it needs check both opname and overload_name then probably it's easier to write them as separate yaml fields? unless opname and overload_name are guaranteed to be concatenated with "." in the codegen script as well.
we could do this in separate PR if needed.

iseeyuan replied:
I think that opname and overload_name are planed to be concatenated with "." in the codegen script. Most of aten ops are already like that, except some jit prim/special ops that miss the overload_name field, but they are going to be updated to "opname.overload_name" format. Is it possible to separate the fields when parsing the yaml file?

iseeyuan replied:
@smessmer Could you confirm the name.overload_name format?

ljk53 commented on binaries/dump_operator_names.cc:
```diff
@@ -0,0 +1,69 @@
+#include <torch/csrc/jit/script/module.h>
```
Note that majority of files under binaries/ have "Copyright (c) 2016-present, Facebook, Inc." license header. Consider following this convention?

Commit:
Update on "dump operator names of a module and its sub-modules."


A binary to dump operator names from a script model and its sub models. 
Usage:
dump_operator_names path/to/script_model.pt path/to/output.yaml

Commit:
Update on "dump operator names of a module and its sub-modules."


A binary to dump operator names from a script model and its sub models. 
Usage:
dump_operator_names path/to/script_model.pt path/to/output.yaml

Commit:
Update on "dump operator names of a module and its sub-modules."


A binary to dump operator names from a script model and its sub models. 
Usage:
dump_operator_names path/to/script_model.pt path/to/output.yaml

Differential Revision: [D18350353](https://our.internmc.facebook.com/intern/diff/D18350353)

pietern commented:
@iseeyuan This raced with #28828 and broke master.

Working on a fix instead of reverting because I think it's an easy fix.

pietern commented:
Turns out the fix is not as quick as I wanted. Reverting this PR.
